File: pascal/tutorials-point/recursion.pas
program recursion;

function factorial(x : integer) : integer;
begin
  if x = 0 then factorial := 1
  else factorial := x * factorial (x - 1);
end;

function fibonacci(x : integer) : integer;
begin
  if x=1 then 
    fibonacci := 0
  else if x=2 then 
    fibonacci := 1
  else
    fibonacci := fibonacci(x - 1) + fibonacci(x - 2);
end;

begin
  writeln('factorial of 6: ', factorial(6));
  writeln('fibonacci of 6: ', fibonacci(6));
end.

File: c/tutorials-point/recursion.c
#include <stdio.h>

/* 
 * classic recusion examples
 * naive factorial and fibonacci
 */

unsigned long long int factorial(unsigned int i) {
  if (i < 2) return 1;
  return i * factorial(i - 1);
}

// very slow
int fibonacci (int i) {
  if (i < 2) return i;
  return fibonacci(i-1) + fibonacci(i-2);
}

int main() {

  for (int i = 1; i < 10; i++)
    printf("number: %3d,   fibonacci: %3d,   factorial: %6lld \n", 
        i, fibonacci(i), factorial(i)); 
  
  return 0;
}

